fabric-commands-v0/src/main/java/net/fabricmc/fabric/mixin/command/MixinCommandManager.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.command;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import net.minecraft.server.command.CommandManager;
+import net.minecraft.server.command.CommandRegistry;
+import net.minecraft.command.AbstractCommand;
+import net.minecraft.command.CommandProvider;
+import net.minecraft.server.MinecraftServer;
+
+import net.fabricmc.fabric.impl.command.FabricCommandRegistryImpl;
+import net.fabricmc.fabric.api.event.server.FabricCommandRegisteredCallback;
+import net.fabricmc.fabric.impl.command.CommandSide;
+
+@Mixin(CommandManager.class)
+public abstract class MixinCommandManager extends CommandRegistry implements CommandProvider {
+	@Inject(method = "<init>", at = @At("TAIL"))
+	public void registerCommands(CallbackInfo info) {
+		FabricCommandRegistryImpl.getCommandMap().forEach((command, side) -> {
+			boolean dedicated = MinecraftServer.getServer().isDedicated();
+
+			if (!(dedicated) && side == CommandSide.CLIENT) {
+				this.registerCommand(command);
+			} else if (dedicated && side == CommandSide.SERVER) {
+				this.registerCommand(command);
+			} else if (side == CommandSide.COMMON) {
+				this.registerCommand(command);
+			}
+
+			FabricCommandRegisteredCallback.EVENT.invoker().onCommandRegistered(MinecraftServer.getServer(), command, side);
```

Comment:
Should this also be called for every vanilla command?

Reply:
people can just mixin into vanilla commands so its not really needed

Reply:
No, it is for the consisitancy, what is the use of this callback anyawys?

Reply:
It's called when a comand  is registered through `FabricCommandRegistry`
The `CommandRegisteredCallback` should go in lifecycle events

Reply:
problem with that. what if the user only has the lifecycle events api?

Reply:
what do you mean by *only has the lifecycle events api*?
unless you plan to include one of the modules, you'd just pull fabric api itself

Reply:
ive seen people use only 1 or 2 modules and just grab them instead of the whole api. unless you find a way to make the lifecycle api depend on the commands api its not a good idea.

Reply:
What is the use case of this event

Reply:
lifecycle events don't need to depend on commands

Reply:
it does if you want to add a command callback in it

Reply:
Events are scattered everywhere in fapi. Those pertaining to the lifecycle of minecraft go in lifecycle events, those that are called by the player go in interaction events, and then there's other scattered events like Item registered and block registered in content registries and packet recieved in networking.